### ISSUE TITLE
Broken font rendering

### ISSUE THREAD

**Issue by moshiur-raj** [flathub/org.chromium.Chromium#238]
![image](https://user-images.githubusercontent.com/36466700/181619357-09129017-1882-4bb5-8f08-d9359f34798c.png)
chromium flatpak in on the left side and native chromium is on the right side. The flatpak version cannot render bengali  font properly. The font in the tab name is also not rendered correctly.

**Comment by tinywrkb** [comment]
Is this just about the tab title and UI elements? I don't see any difference in web content text. If this is the case, then please make this clear.  
My guess is that your system has a font suite like Noto or Adobe Source Fonts installed with Fontconfig configs to prefer its Families for Sans, Serief, and Monospace font, which is why the system app use it for UI elements.  
For example, this is for preferring Noto Sans as Sans font.
```
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
  <alias>
    <family>Noto Sans</family>
    <default>
      <family>sans-serif</family>
    </default>
  </alias>
</fontconfig>
``` 
While Flatpak makes system fonts available to sandboxed applications, it does not apply system Fontconfig configs, so fonts are there, but they are not being used.  

A possible workaround is to copy the Fontconfig configs into `~/.var/app/org.chromium/Chromium/config/fontconfig/fonts.conf`.  
This could actually be set more globally, but this can go wrong (fontconfig is finicky), so I don't recommend that unless you know what you're doing.  
Please see `man:fonts-conf(5)` for more details. There's also a good [Arch Linux Wiki article](https://wiki.archlinux.org/title/Font_configuration), and [examples](https://wiki.archlinux.org/title/Font_configuration/Examples).

**Comment by moshiur-raj** [comment]
The web content is different. Ligatures are broken.
![image](https://user-images.githubusercontent.com/36466700/181632627-622bc99f-59a6-4b81-bf8d-468d76f0ba8b.png)

**Comment by tinywrkb** [comment]
For web contents, try comparing the fonts used for rendering.  
Open devtools (ctrl+shift+i), select the Elements tab (default). On the right pane, you have Computed tab next to Styles, press on it. Now move the cursor over the text in question, open the context menu (mouse right click) and press on Inspect, and then go back to the Computed tab, scroll down, and inspect the Rendered Fonts list.

**Comment by moshiur-raj** [comment]
The flatpak is using freesans. The native version is using noto sans and liberation sans.

**Comment by tinywrkb** [comment]
So we're back to what I said at my first post, this is an issue of system Fontconfig configs not being picked by Flatpak.

Try something like this

`~/.var/app/org.chromium/Chromium/config/fontconfig/fonts.conf`
```
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
 <alias>
   <family>sans-serif</family>
   <prefer>
     <family>Noto Sans</family>
     <family>Liberation Sans</family>
     <family>Noto Color Emoji</family>
     <family>Noto Emoji</family>
   </prefer>
 </alias>

 <alias>
   <family>serif</family>
   <prefer>
     <family>Noto Serif</family>
     <family>Liberation Serif</family>
     <family>Noto Color Emoji</family>
     <family>Noto Emoji</family>
   </prefer>
 </alias>

 <alias>
  <family>monospace</family>
  <prefer>
    <family>Noto Sans Mono</family>
     <family>Liberation Mono</family>
    <family>Noto Color Emoji</family>
    <family>Noto Emoji</family>
   </prefer>
 </alias>

</fontconfig>
```

**Comment by moshiur-raj** [comment]
Is the path correct? Shouldn't it be `org.chromium.Chromium` instead of `org.chromium/Chromium` ? I used `org.chromium.Chromium` and the fonts in the web content is fixed. But the fonts in ui elements still have problems (broken ligature).

**Comment by tinywrkb** [comment]
> Is the path correct? Shouldn't it be org.chromium.Chromium instead of org.chromium/Chromium

Correct, sorry about that.

> I used org.chromium.Chromium and the fonts in the web content is fixed. But the fonts in ui elements still have problems (broken ligature).

Maybe Chromium is matching the font used by GTK3. Try setting GTK3 font to Noto Sans or sans.  
You can create `~/.var/app/org.chromium.Chromium/gtk-3.0/settings.ini` as described in [this Arch Wiki article](https://wiki.archlinux.org/title/GTK#Basic_theme_configuration). If this works, then we should actually fix the app here, and have it pick up the global `~/.config/gtk-3.0/settings.ini`.  
Actually, if this works, and you're using a desktop environment, then I suspect that you didn't set your system font.  
Alternatively, you can also fix this via `fonts.conf`. Just add something like this
```
 <alias>
  <family>Cantarell</family>
  <prefer>
    <family>Noto Sans</family>
     <family>Liberation Sans</family>
  </prefer>
 </alias>
```
Note here that Cantarell is the default GTK3 font.  
And there's another way to fix this
```
$ flatpak run --command=gesttings org.chromium.Chromium get org.gnome.desktop.interface font-name
$ flatpak run --command=gesttings org.chromium.Chromium set org.gnome.desktop.interface font-name "Noto Sans"
```

**Comment by moshiur-raj** [comment]
I tried all three methods. Did not work.

**Comment by moshiur-raj** [comment]
You can visit [this link](https://www.prothomalo.com/) to test things yourself.
![image](https://user-images.githubusercontent.com/36466700/181755062-ea9e9a95-bb98-4586-ad08-da7440f034ac.png)
![image](https://user-images.githubusercontent.com/36466700/181755164-a362709c-1424-4102-9643-128b908d216f.png)
Notice how the ligatures are broken in the tab name for chromium. The fonts are rendered correctly in firefox. Both are flatpaks.

**Comment by tinywrkb** [comment]
> I tried all three methods. Did not work.

The GTK3 setting.ini method seems to work when explicitly choosing the Bengali font.

`~/.var/app/org.chromium.Chromium/config/gtk-3.0/settings.ini`
```
[Settings]
gtk-font-name = Noto Sans Bengali 11
```

I really don't like this method, lots of users are polyglot, so this type of solution is not adequate. The Fontconfig alias should have worked, and make it possible to have priorities and fallbacks.

**Comment by tinywrkb** [comment]
And I had a typo in the gsettings command (`gesttings` instead of `gsettings`).

**Comment by moshiur-raj** [comment]
> And I had a typo in the gsettings command (gesttings instead of gsettings).

Yeah, you also forgot to add `config` after `org.chromium.Chromium` in the directory for `settings.ini`.

**Comment by moshiur-raj** [comment]
> The GTK3 setting.ini method seems to work when explicitly choosing the Bengali font.

Adding `Noto Sans Bengali` in the fontconfig solves the problem too.

**Comment by tinywrkb** [comment]
> Adding `Noto Sans Bengali` in the fontconfig solves the problem too.

Where? It didn't work for me. With the alias for Cantarell (default GTK3 font) or sans-serif alias?

If you set alias for Cantarell, then you might also want to set alias for Source Code Pro which is the default GTK3 Monospace font. Though I don't see a Bengali Mono font.

**Comment by moshiur-raj** [comment]
This is the `fonts.conf` I am using:
```
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
 <alias>
   <family>sans-serif</family>
   <prefer>
     <family>Noto Sans</family>
     <family>Noto Sans Bengali</family>
     <family>Liberation Sans</family>
     <family>Noto Color Emoji</family>
     <family>Noto Emoji</family>
   </prefer>
 </alias>

 <alias>
   <family>serif</family>
   <prefer>
     <family>Noto Serif</family>
     <family>Noto Serif Bengali</family>
     <family>Liberation Serif</family>
     <family>Noto Color Emoji</family>
     <family>Noto Emoji</family>
   </prefer>
 </alias>

 <alias>
  <family>monospace</family>
  <prefer>
    <family>Noto Sans Mono</family>
     <family>Liberation Mono</family>
    <family>Noto Color Emoji</family>
    <family>Noto Emoji</family>
   </prefer>
 </alias>

</fontconfig>
```

**Comment by moshiur-raj** [comment]
This config works for Signal and Tutanota as well. I think this will work with all electron based apps.
Manually mentioning the font is a pain though. Can't it be automated? I wonder what firefox is doing different. It works right out of the box for firefox.

**Comment by tinywrkb** [comment]
> Manually mentioning the font is a pain though. Can't it be automated?

Try moving `fonts.conf` to `~/.config/fontconfig/fonts.conf` and run Chromium with `$ flatpak run --filesystem=xdg-config/fontconfig org.chromium.Chromium`.  
If this works, then you can add this filesystem override globally with Flatseal or `$ flatpak overrride --user --filesystem=xdg-config/fontconfig`, or per app with `$ flatpak overrride --user --filesystem=xdg-config/fontconfig org.chromium.Chromium`, etc.  

I don't like this at all, but I don't see any other solution. This is a Flatpak bug that should be fixed. Flatpak should at least source the user's `fonts.conf` config, and as long as there's no other way to set Fontcnofig, desktop environments should set `fonts.conf` or dedicated configs in `fontconfig/conf.d`.

**Comment by tinywrkb** [comment]
> I wonder what firefox is doing different. It works right out of the box for firefox.

I haven't looked into the code, so I'm just guessing here, but I suspect (with a good reason, as I hit related issues to the lib) that Firefox is using Pango.

**Comment by moshiur-raj** [comment]
Hmm, I think I'm going to keep using separate `fonts.conf` for each app. I'd rather not mess with global options. I hope this gets fixed soon. Thank you for your help.

**Comment by tinywrkb** [comment]
@moshiur-raj correct me if I'm wrong, but from I'm seeing ligatures were broken in UI elements also for native Chromium app. So I wonder if we have here a generic / non-Flatpak issue with Noto Fonts. This need to be figured out, as I'm not convinced that the user should be expected to add `Noto Sans Bengali` to `fonts.conf.

I think that the missing support for system and user Fontconfig configs makes the case for adding Noto Fonts with its superb Language coverage to the Flatpak runtime, but knowing the previous discussion about this, I'm not optimistic.  
In my system, I have a full Noto Fonts suite, including the CJK fonts, which sums up to around 200MiB, and that's due to usage of variable fonts, where available.  
Even if we say that font files are totally different in different runtime versions, so no dedup, we're talking about 600MiB if we have 3 maintained runtime versions, and about half of this is the user has storage compression enabled (must have on non-mechanical drive to limit write).

**Comment by moshiur-raj** [comment]
@tinywrkb ligatures aren't broken in the native chromium app. I have used the native version of chromium for years, never encountered broken ligatures. It only became a problem after switching to the flatpak version.

**Comment by moshiur-raj** [comment]
@tinywrkb  I think it is a problem with electron / chromium based apps.
![image](https://user-images.githubusercontent.com/36466700/181838505-b0f5ec11-788c-4098-b2c2-d31a8514781d.png)
notice the difference between the ibus preview and the text box in signal. This is no longer an issue if I edit the fontconfig. I also opened an [issue](https://github.com/flathub/org.signal.Signal/issues/330) in `org.signal.Signal`.

**Comment by tinywrkb** [comment]
> @tinywrkb ligatures aren't broken in the native chromium app. I have used the native version of chromium for years, never encountered broken ligatures. It only became a problem after switching to the flatpak version.

Alright, thanks. I would like to see this working in Flatpak without adding the regional font `Noto Sans Bengali` to `font.conf`. I think that we can then try to make the case for inclusion of Noto Fonts in the Flatpak runtime. I don't know if Flatpak sourcing system Fontconfig configs is something that will be supported any time soon.  
This kind of user experience is unacceptable.  
BTW, what distro and version are you using?

**Comment by moshiur-raj** [comment]
I am using archlinux and the flatpak version is listed as 1:1.12.7-1
Chromium and other flatpak libraries are up to date.

**Comment by tinywrkb** [comment]
Oh! A fellow Arch Linux user. So I'm using the same system Fontconfig configs for Noto Fonts as yours.

**Comment by Erick555** [comment]
> I also opened an [issue](https://github.com/flathub/org.signal.Signal/issues/330) in `org.signal.Signal`.

Please don't open issues for specific apps elsewhere as this is generic flatpak problem.

The noto size is big blocker for adding it in runtime. The fonts already should be available from host so I wonder if adding just configs from noto would work

**Comment by moshiur-raj** [comment]
@Erick555 Sorry about that. I actually opened the issue in Signal first then realized it is a problem with chromium as well. Didn't know it was a flatpak problem at first.

**Comment by tinywrkb** [comment]
> The fonts already should be available from host so I wonder if adding just configs from noto would work

Maybe, maybe not needed.

I looked at this again today, and I suggest waiting for the runtime to be updated to Freedesktop 22.08.  
With Fontconfig 2.14 we have Noto Fonts families included in the default configs.
I tried to figured out if this change solves the problem of web content's ligatures without needing user-defined `fonts.conf`, but I don't have something to test with, as the example `www.prothomalo.com` brings its own font.

The second problem, ligatures in UI elements, seems to be solved when properly a Bengali locale.  
Setting `LANG`, or `LANGUAGE`, or even `FC_LANG` to `bn_BD.UTF-8` brings the desired result.  
Actually, I couldn't get this working without setting the locale to `bn_BD.UTF-8`, so I'm not sure what's going on.  
The fact that Fontconfig's utils are crashing in Freedesktop 22.08beta doesn't help.  
I'm not sure the requirement to set a Bengali locale is an extra step to what's needed to make these ligatures work compared to host Chromium app, but this creates a problem if the user also needs Noto CJK and depends on the locale for correct regional character selection (Chinese Simplified, Traditional, ...).

I tried injecting configs into the runtime and seeing if the situation is improved, but no success so far.  
Also, I'm not the right person to test this, as I can't read Bengali.

TLDR: Wait for Freedesktop 22.08 (almost here) and the runtime bump to this app, test again, and if still deviates from host Chromium, then open a bug report at [the Freedesktop runtime bugtracker](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues).

**Comment by sajjadkabirjoy** [comment]
After spending a couple of days trying to solve the issue, here's what worked for me. I'm on Debian Unstable.

Flatpak Chromium (and possibly other Flatpak apps) rely on the GNU FreeFont (FreeSans, FreeSerif and FreeMono) for rendering non-Latin characters. While FreeFont might be the only thing out there with a good portion of the entire Unicode charset crammed into a single font file, it fails quite miserably while rendering ligatures in some Asian writing systems, Bangla (Bengali) being one of them. So FreeFont needs to be disabled. This can be achieved by placing the following in `~/.var/app/org.chromium.Chromium/config/fontconfig/fonts.conf`: 
```
<selectfont>
     <rejectfont>
         <pattern>
             <patelt name="family" >
                 <string>FreeSans</string>
             </patelt>
         </pattern>
     </rejectfont>
 </selectfont>
 <selectfont>
     <rejectfont>
         <pattern>
             <patelt name="family" >
                 <string>FreeSerif</string>
             </patelt>
         </pattern>
     </rejectfont>
 </selectfont>
 <selectfont>
     <rejectfont>
         <pattern>
             <patelt name="family" >
                 <string>FreeMono</string>
             </patelt>
         </pattern>
     </rejectfont>
 </selectfont>
```
I have a bunch of other fonts disabled as well (the Liberation fonts, for instance) and set substitutes for fonts like Arial, Times, Courier etc. to the effect that they are replaced by the respective Noto sans, serif or monospace fonts. Et voilà! Attaching some screenshots to reflect the changes.
![Screenshot from 2023-09-28 17-18-07](https://github.com/flathub/org.chromium.Chromium/assets/122977683/391101a6-b8b2-46c2-9111-fd383fd86f6a)
![Screenshot from 2023-09-28 17-20-58](https://github.com/flathub/org.chromium.Chromium/assets/122977683/434759af-ff81-4d30-a9c5-bb0c5ec6d085)

Hope this helps other users suffering from the same issue.

**Comment by codayon** [comment]
This is probably related to https://github.com/flatpak/flatpak/issues/1563 and many others. Perhaps it's an issue with Flatpak itself and is not inherently connected to Chromium nor [GTK-based](https://github.com/flatpak/flatpak/issues/5730) applications.

**Comment by ratul0547** [comment]
Same issue still persists in 2026.
